runtime/bcutil/cfreader.c

```diff
@@ -2642,9 +2642,13 @@ j9bcutil_readClassFileBytes(J9PortLibrary *portLib,
 
 	/* Make sure the structure and static verification below uses the class file version
 	 * number. VM version is maintained in vmVersionShifted.
+	 * Note: ignore the conversion of version to avoid the version check at checkBytecodeStructure
+	 * in the case of Unsafe.defineClass().
 	 */
-	flags &= ~BCT_MajorClassFileVersionMask;
-	flags |= ((UDATA) classfile->majorVersion) << BCT_MajorClassFileVersionMaskShift;
+	if (J9_ARE_NO_BITS_SET(findClassFlags, J9_FINDCLASS_FLAG_UNSAFE)) {
```

Comment:
Rather than making this change, which is quite distant from the source of the issue, could staticverify.c `checkBytecodeStructure` be changed instead?

Something like:
```c
if (info->tag != CFR_CONSTANT_Methodref) {
	if (((flags & BCT_MajorClassFileVersionMask) >= BCT_Java8MajorVersionShifted)
	&& (bc != CFR_BC_invokevirtual) && (info->tag == CFR_CONSTANT_InterfaceMethodref)
	) {
		/* JVMS 4.9.1 Static Constraints:
		 * The indexbyte operands of each invokespecial and invokestatic instruction must represent
		 * a valid index into the constant_pool table. The constant pool entry referenced by that
		 * index must be either of type CONSTANT_Methodref or of type CONSTANT_InterfaceMethodref.
		 */
		 /* Valid - take no action */
	} else {
		if (flags & BCT_Unsafe) {
			// do any version specific checks here for 8 / 11 / etc that may be needed
			// requires adding a new BCT_Unsafe flag
		} else {
			errorType = J9NLS_CFR_ERR_BC_NOT_METHODREF__ID;
			/* Jazz 82615: Set the constant pool index to show up in the error message framework */
			errorDataIndex = index;
			goto _verifyError;
		}
	}
}
```

Reply:
Technically it should be fine to move this check to `checkBytecodeStructure()` but we still need to check the bytecode & `CFR_CONSTANT_InterfaceMethodref` in the case of unsafe to restrict the usage of unsafe to these cases. e.g.

```
if ( (J9_ARE_ANY_BITS_SET(flags & BCT_Unsafe) || ((flags & BCT_MajorClassFileVersionMask) >= BCT_Java8MajorVersionShifted))
	&& (bc != CFR_BC_invokevirtual) && (info->tag == CFR_CONSTANT_InterfaceMethodref)
	) {
		/* JVMS 4.9.1 Static Constraints:
		 * The indexbyte operands of each invokespecial and invokestatic instruction must represent
		 * a valid index into the constant_pool table. The constant pool entry referenced by that
		 * index must be either of type CONSTANT_Methodref or of type CONSTANT_InterfaceMethodref.
		 */
		 /* Valid - take no action */
	}  else {  ...}
```
or 
```
BOOLEAN checkFlag = J9_ARE_ANY_BITS_SET(flags & BCT_Unsafe || ((flags & BCT_MajorClassFileVersionMask) >= BCT_Java8MajorVersionShifted));

if ( checkFlag && (bc != CFR_BC_invokevirtual) && (info->tag == CFR_CONSTANT_InterfaceMethodref)
	) {
		/* JVMS 4.9.1 Static Constraints:
		 * The indexbyte operands of each invokespecial and invokestatic instruction must represent
		 * a valid index into the constant_pool table. The constant pool entry referenced by that
		 * index must be either of type CONSTANT_Methodref or of type CONSTANT_InterfaceMethodref.
		 */
		 /* Valid - take no action */
	} else {   ...}
```

Reply:
The latest changes above didn't work correctly as the failure occurred again at `jsr335InterfacePrivateMethod`
```
java.lang.VerifyError: JVMCFRE116 invoke bytecode must reference a Methodref; class=jdk/internal/reflect/GeneratedMethodAccessor18, method=invoke(Ljava/lang/Object;[Ljava/lang/Object;)Ljava/lang/Object;, pc=20
Exception Details:
  Location:
    jdk/internal/reflect/GeneratedMethodAccessor18.invoke(Ljava/lang/Object;[Ljava/lang/Object;)Ljava/lang/Object; @20: JBinvokestatic
  Reason:
    Wrong type at the index 10 of constant pool
  Exception Handler Table:
    bci [0, 20] => handler: 24
    bci [0, 20] => handler: 24
    bci [20, 23] => handler: 36
	at java.base/jdk.internal.misc.Unsafe.defineClass(Unsafe.java:1510)
	at java.base/jdk.internal.reflect.ClassDefiner.defineClass(ClassDefiner.java:63)
	at java.base/jdk.internal.reflect.MethodAccessorGenerator$1.run(MethodAccessorGenerator.java:400)
	at java.base/jdk.internal.reflect.MethodAccessorGenerator$1.run(MethodAccessorGenerator.java:394)
	at java.base/java.security.AccessController.doPrivileged(AccessController.java:678)

```
Need to double-check to figure out what happened to the updated code.

Reply:
Just figured that `flags` was not set up with `BCT_Unsafe` before use. So updated to the code as follows in j9bcutil_readClassFileBytes at cfreader.c
```
	if (J9_ARE_ANY_BITS_SET(findClassFlags, J9_FINDCLASS_FLAG_UNSAFE)) {
		flags |= BCT_Unsafe;  <--------
	}
```
Will launch personal builds again to see how it goes in this way.